### PR TITLE
Modify resize_token_embeddings to ensure output type is same as input

### PR DESCRIPTION
**Pull request by bayllama**
# What does this PR do?

Modified resize_token_embeddings to return the same class that is passed as input to it. Today, even if a custom embedding class is passed, resize_token_embeddings converts it to a nn.Embedding but this commit makes sure that does not happen and the custom embedding class is returned.

Fixes # (31835)
[](https://github.com/huggingface/transformers/issues/31835)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@zucchini-nlp

### CODE REVIEW

**Review by zucchini-nlp — COMMENTED** [review]
Great work @bayllama !

One thing to note is that the newly created Embedding will not have the same embed scale as the old one, because we don't pass `embed_scale` at creation and the default is 1.0.

What if we don't rely on what type of embedding class is being used and modify the weights of the `old_embedding` in-place and return it as `new_embeddings`. Something like this added at the end before returning

```
old_embeddings.weight.data = new_embeddings.weight.data
return old_embeddings
```

@amyeroberts WDYT of this idea? It doesn't break BC and makes `ModelScaledEmbedding` happy

**Comment by bayllama** [comment]
@zucchini-nlp 
That's a good idea. I have made the change you recommended.

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -2096,7 +2096,8 @@ def _get_resized_embeddings(
         else:
             new_embeddings.weight.data[:n, :] = old_embeddings.weight.data[:n, :]
 
-        return new_embeddings
+        old_embeddings.weight.data = new_embeddings.weight.data
```
Comment (amyeroberts) [inline src/transformers/modeling_utils.py:2099]: Could we add a comment above to explain why we're doing this? Even just: 

```suggestion
        # c.f. https://github.com/huggingface/transformers/pull/31979
        old_embeddings.weight.data = new_embeddings.weight.data
```

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing and nice suggestion @zucchini-nlp!

**Comment by bayllama** [comment]
@zucchini-nlp @amyeroberts Added the comment describing the change

**Review by zucchini-nlp — APPROVED** [review]
Perfect, thanks for working on this! I'll make sure that tests are passing and it can be merged later today

**Comment by zucchini-nlp** [comment]
@bayllama the tests are failing because we pass `device` and `dtype` to the Embedding Class when resizing, but the custom classes don't accept any kwargs. Can you add `**kwargs` plz to init in the correct device and dtype?

You should be able to verify all tests are passing by this command :)
```bash
pytest -k test_resize_token tests/models/
```

**Comment by bayllama** [comment]
@zucchini-nlp 

When I run the pytest that you suggested I don't see the error that you have mentioned above. We don't set the device and dtype for the Custom Embedding Class at all right?

```
new_embeddings = nn.Embedding(
            new_num_tokens,
            old_embedding_dim,
            device=old_embeddings.weight.device,
            dtype=old_embeddings.weight.dtype,
        )
```

We only do it for the nn.Embedding and the replace the weight.data in the old embeddings which is already in the correct device and has the right dtype.

However I do see another issue which is in hte resize_token_embeddings, this attribute "text_config" is being modified and I haven't accounted for that in my previous commit. I will work on a fix for this. However can you please give me more details on the dtype and device issue, because I don't see it in my set of tests.

**Comment by zucchini-nlp** [comment]
@bayllama my bad, I was running in the a different branch. You're right, some VLMs are failing, as I see it should be fixed with one line after we swapped weights! Let me know when it's fixed :)

**Comment by bayllama** [comment]
@zucchini-nlp Changing the shape of the old_embeddings would fix this, however shape attribute is not writable in torch, hence we cannot do something like this,
```
old_embeddings.weight.shape = new_embeddings.weight.shape
```

Hence I am thinking of what would be the best way to do this. If nothing works out I may need to go to the method I was following in my first PR where I check the type of input embedding and create a new object and return it.

**Comment by zucchini-nlp** [comment]
@bayllama I see, but we can also change the `num_embeddings` attribute which is writable.
Like `old_embeddings.num_embeddings = new_embeddings.weight.data.shape[0]`

**Comment by bayllama** [comment]
@zucchini-nlp  In addition to the above I found a couple of more things,

1) The padding_idx must be updated to None if the number of tokens in the new embeddings is smaller than the padding_idx. 

2) The model Lxmert has a bias term that has to be updated similar to the final_logits_bias in mbart. This if not updated would cause some failures in test cases. This problem has been there from before and I believe was not addressed.

In addition to what we already discussed about, I have made these changes in this commit as well.  All of the test cases are passing for me now.

**Review by Khaliq88 — APPROVED** [review]

**Comment by bayllama** [comment]
@zucchini-nlp I am not sure what this tests_hub failure is. Could you please help out

**Review by zucchini-nlp — COMMENTED** [review]
Great! Some final nits and we need to merge `main` as a final step, then we're good to go. 

The failing tests are not related to this PR and should be resolved by rerunning them.

**Comment by bayllama** [comment]
@zucchini-nlp Made the changes that you have recommended. Please take a look. Thanks!

**Comment by zucchini-nlp** [comment]
@bayllama thanks, everything looks good! Rerunning the tests didn't help, so can you merge `main` in case it was already fixed there

**Comment by zucchini-nlp** [comment]
@amyeroberts can you merge this plz? Unrelated hub failures, from internal slack seems like we're not the only ones seeing it

**Comment by bayllama** [comment]
@zucchini-nlp @amyeroberts  Seems like the tests went through after rebasing. Let me know if anything else is required here.

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -2129,7 +2129,19 @@ def _get_resized_embeddings(
...
+        # This ensures correct functionality when a Custom Embedding class is passed as input.
+        # The input and output embedding types remain consistent. (c.f. https://github.com/huggingface/transformers/pull/31979)
+        old_embeddings.weight.data = new_embeddings.weight.data
+        old_embeddings.num_embeddings = new_embeddings.weight.data.shape[0]
+
+        # If the new number of tokens is smaller than the original `padding_idx`, the `padding_idx`
+        # will be set to `None` in the resized embeddings.
+        if old_embeddings.padding_idx is not None:
+            if (new_num_tokens - 1) < old_embeddings.padding_idx:
+                old_embeddings.padding_idx = None
```
Comment (amyeroberts) [inline src/transformers/modeling_utils.py:2142]: nit - no need to have this nested if here 

```suggestion
        if old_embeddings.padding_idx is not None and (new_num_tokens - 1) < old_embeddings.padding_idx:
            old_embeddings.padding_idx = None
```

**Review by amyeroberts — COMMENTED** [review]
Looks great - thanks @bayllama! 

One last thing I forgot to ask for before merge is to add a test. There's already tests for resizing embeddings, so extending those to check the returned type should be enough

Could you also update the title so it's no longer truncated?

**Comment by bayllama** [comment]
@amyeroberts @zucchini-nlp Added the test to make sure the correct type is returned. Also made the other changes suggested.

**Comment by bayllama** [comment]
@zucchini-nlp @amyeroberts I believe zucchini-nlp already pushed in a fix for this. Let me know if I need to do something.

**Pull request merged**